src/callbacks_step/amr_dg3d.jl

```diff
@@ -69,6 +69,109 @@ function refine!(u_ode::AbstractVector, adaptor,
+        GC.@preserve old_inverse_jacobian begin # OBS! If we don't GC.@preserve old_inverse_jacobian, it might be GC'ed
+            old_u = wrap_array(old_u_ode, mesh, equations, dg, cache)
+
+            # Loop over all elements in old container and scale the old solution by the Jacobian
+            # prior to projection
+            for old_element in 1:old_n_elements
+                for k in eachnode(dg), j in eachnode(dg), i in eachnode(dg)
+                    for v in eachvariable(equations)
+                        old_u[v, i, j, k, old_element] = (old_u[v, i, j, k,
+                                                                old_element] /
+                                                          old_inverse_jacobian[i, j, k,
+                                                                               old_element])
+                    end
+                end
+            end
+
+            reinitialize_containers!(mesh, equations, dg, cache)
+
+            resize!(u_ode,
+                    nvariables(equations) * nnodes(dg)^ndims(mesh) *
+                    nelements(dg, cache))
+            u = wrap_array(u_ode, mesh, equations, dg, cache)
+
+            # Loop over all elements in old container and either copy them or refine them
+            u_tmp1 = Array{eltype(u), 4}(undef, nvariables(equations), nnodes(dg),
+                                         nnodes(dg), nnodes(dg))
+            u_tmp2 = Array{eltype(u), 4}(undef, nvariables(equations), nnodes(dg),
+                                         nnodes(dg), nnodes(dg))
+            element_id = 1
+            for old_element_id in 1:old_n_elements
+                if needs_refinement[old_element_id]
+                    # Refine element and store solution directly in new data structure
+                    refine_element!(u, element_id, old_u, old_element_id,
+                                    adaptor, equations, dg, u_tmp1, u_tmp2)
+                    # Before `element_id` is incremented, divide by the new Jacobians on each
+                    # child element and save the result
+                    for m in 0:7 # loop over the children
+                        for k in eachnode(dg), j in eachnode(dg), i in eachnode(dg)
+                            for v in eachvariable(equations)
+                                u[v, i, j, k, element_id + m] *= (0.125 *
+                                                                  cache.elements.inverse_jacobian[i,
+                                                                                                  j,
+                                                                                                  k,
+                                                                                                  element_id + m])
+                            end
+                        end
+                    end
+                    element_id += 2^ndims(mesh)
```

Comment:
Same here, potential confusion with line 155

Reply:
Done in 0eea3769018e18969dcff326e1152811a16e155e